### PR TITLE
`sage.features.topcom`

### PR DESCRIPTION
**Pull request by mkoeppe**
- Part of #35856

We also change how the SPKG index in the reference manual is generated.
`index.rst` is now a static file that uses `.. include` directives to read in generated files `index_standard.rst`, `index_optional.rst`, etc.

### :memo: Checklist

- [x] The title is concise and informative.
- [ ] The description explains in detail what this PR is about.
- [x] I have linked a relevant issue or discussion.
- [ ] I have created tests covering the changes.
- [ ] I have updated the documentation and checked the documentation preview.

### CODE REVIEW

**Comment by kwankyu** [comment]
This needs to be updated: https://github.com/sagemath/sage/blob/15fcf52fbd6a72f0d0f226c02fe1a74b21bbfaf5/src/doc/bootstrap#L163

(btw, I think this list should be automatically generated and split from spkgs list and moved to a new location.)

**Comment by mkoeppe** [comment]
Shall we use a sphinx `.. include` directive for that?

**Comment by kwankyu** [comment]
I don't know. Anyway, this is for a new PR. I may work on it later.

**Comment by mkoeppe** [comment]
Here's an attempt. The list of features is not autogenerated (we don't do that for any Python modules); but at least it's now in an ordinary editable index file.

**Review by kwankyu — APPROVED** [review]
Thanks.

LGTM.

**Comment by mkoeppe** [comment]
Thank you!

**Comment by vbraun** [comment]
I'm getting 
```
[sagemath_doc_html-none] [spkg-install] sage --docbuild --no-pdf-links reference/modules inventory --no-prune-empty-dirs
[sagemath_doc_html-none] [spkg-install] [spkg     ] The inventory file is in ../../local/share/doc/sage/inventory/en/reference/spkg.
[sagemath_doc_html-none] [spkg-install] Error building the documentation.
[sagemath_doc_html-none] [spkg-install] Traceback (most recent call last):
[sagemath_doc_html-none] [spkg-install]   File "<frozen runpy>", line 198, in _run_module_as_main
[sagemath_doc_html-none] [spkg-install]   File "<frozen runpy>", line 88, in _run_code
[sagemath_doc_html-none] [spkg-install]   File "/home/release/Sage/src/sage_docbuild/__main__.py", line 514, in <module>
[sagemath_doc_html-none] [spkg-install]     sys.exit(main())
[sagemath_doc_html-none] [spkg-install]              ^^^^^^
[sagemath_doc_html-none] [spkg-install]   File "/home/release/Sage/src/sage_docbuild/__main__.py", line 510, in main
[sagemath_doc_html-none] [spkg-install]     builder()
[sagemath_doc_html-none] [spkg-install]   File "/home/release/Sage/src/sage_docbuild/builders.py", line 823, in _wrapper
[sagemath_doc_html-none] [spkg-install]     getattr(DocBuilder, build_type)(self, *args, **kwds)
[sagemath_doc_html-none] [spkg-install]   File "/home/release/Sage/src/sage_docbuild/builders.py", line 163, in f
[sagemath_doc_html-none] [spkg-install]     runsphinx()
[sagemath_doc_html-none] [spkg-install]   File "/home/release/Sage/src/sage_docbuild/sphinxbuild.py", line 319, in runsphinx
[sagemath_doc_html-none] [spkg-install]     sys.stderr.raise_errors()
[sagemath_doc_html-none] [spkg-install]   File "/home/release/Sage/src/sage_docbuild/sphinxbuild.py", line 255, in raise_errors
[sagemath_doc_html-none] [spkg-install]     raise OSError(self._error)
[sagemath_doc_html-none] [spkg-install] OSError: /home/release/Sage/src/doc/en/reference/spkg/index.rst:28: WARNING: toctree contains reference to nonexisting document 'sage/features'
```

**Comment by mkoeppe** [comment]
@vbraun Is that in an incremental build? This looks like #37786.

Anyway I have pushed a possible fix

**Comment by vbraun** [comment]
Yes, incremental

**Comment by vbraun** [comment]
Still fails. Try
```
./bootstrap
cd src/doc
make doc-inventory--reference-spkg     # fail
make doc-inventory--reference-spkg     # succeeds
```

**Comment by mkoeppe** [comment]
I can't reproduce a failure

**Comment by vbraun** [comment]
I'm getting 
```
[sagemath_doc_html-none] [spkg-install] Error building the documentation.
[sagemath_doc_html-none] [spkg-install] Traceback (most recent call last):
[sagemath_doc_html-none] [spkg-install]   File "<frozen runpy>", line 198, in _run_module_as_main
[sagemath_doc_html-none] [spkg-install]   File "<frozen runpy>", line 88, in _run_code
[sagemath_doc_html-none] [spkg-install]   File "/home/release/Sage/src/sage_docbuild/__main__.py", line 530, in <module>
[sagemath_doc_html-none] [spkg-install]     sys.exit(main())
[sagemath_doc_html-none] [spkg-install]              ^^^^^^
[sagemath_doc_html-none] [spkg-install]   File "/home/release/Sage/src/sage_docbuild/__main__.py", line 526, in main
[sagemath_doc_html-none] [spkg-install]     builder()
[sagemath_doc_html-none] [spkg-install]   File "/home/release/Sage/src/sage_docbuild/builders.py", line 823, in _wrapper
[sagemath_doc_html-none] [spkg-install]     getattr(DocBuilder, build_type)(self, *args, **kwds)
[sagemath_doc_html-none] [spkg-install]   File "/home/release/Sage/src/sage_docbuild/builders.py", line 163, in f
[sagemath_doc_html-none] [spkg-install]     runsphinx()
[sagemath_doc_html-none] [spkg-install]   File "/home/release/Sage/src/sage_docbuild/sphinxbuild.py", line 319, in runsphinx
[sagemath_doc_html-none] [spkg-install]     sys.stderr.raise_errors()
[sagemath_doc_html-none] [spkg-install]   File "/home/release/Sage/src/sage_docbuild/sphinxbuild.py", line 255, in raise_errors
[sagemath_doc_html-none] [spkg-install]     raise OSError(self._error)
[sagemath_doc_html-none] [spkg-install] OSError: /home/release/Sage/src/doc/en/reference/spkg/index.rst:28: WARNING: toctree contains reference to nonexisting document 'sage/features'
```

**Comment by vbraun** [comment]
Looks like this is a parallel build race, works when building in a single thread. But `make doc-clean && make -j10 doc-html` errors out reliably for me. One of the parallel tasks prints a line
```
[sagemath_doc_html-none] [spkg-install] Deleting empty directory /home/release/Sage/src/doc/en/reference/spkg/sage/features
```
Seems like the doc-inventory--reference-spkg target conflicts

**Comment by mkoeppe** [comment]
Thanks for the details. I'll try to repro

**Comment by kwankyu** [comment]
@vbraun What is your system?

**Comment by kwankyu** [comment]
@vbraun Would you try #38185?

**Comment by kwankyu** [comment]
@vbraun Does #38185 work well with you? Then it needs review.

**Comment by vbraun** [comment]
Worked for me

**Comment by kwankyu** [comment]
Great! Thanks.

**Comment by vbraun** [comment]
Actually no, still happens upon further testing

**Comment by kwankyu** [comment]
It is hard to "fix" an invisible issue. So what is your system and reproducing procedure?

**Comment by vbraun** [comment]
The following steps reproducably cause the failure for me:
```
$ git clean -f -d -x ./src/doc/en/reference/spkg && ./bootstrap && touch src/doc/en/reference/spkg/index.rst  && ./sage --docbuild --no-pdf-links reference/spkg inventory --no-prune-empty-dirs
[...]
OSError: /home/release/Sage/src/doc/en/reference/spkg/index.rst:28: WARNING: toctree contains reference to nonexisting document 'sage/features'
```
Running the docbuild command a second time succeeds, then
```
$ ./sage --docbuild --no-pdf-links reference/spkg inventory --no-prune-empty-dirs
[spkg     ] updating environment: 31 added, 1 changed, 0 removed
[spkg     ] The inventory file is in local/share/doc/sage/inventory/en/reference/spkg.
```

**Comment by kwankyu** [comment]
Don't you get the same failure if you replace 'spkg' with 'coding' everywhere?

**Comment by kwankyu** [comment]
After more experiments, I now realized there is nothing strange in the above experiments. We are breaking assumptions by `git clean -f -d -x ./src/doc/en/reference/spkg` (or `coding`). No fault on the doc build mechanism.

The doc build starts initially with the state made by `make doc-clean` (cleaning `src/doc/sage`) and `make doc-uninstall` (removing `local/share/doc`). This should work
```
$ make doc-uninstall && git clean -f -d -x ./src/doc/en/reference/spkg && ./bootstrap && touch src/doc/en/reference/spkg/index.rst  && ./sage --docbuild --no-pdf-links reference/spkg inventory --no-prune-empty-dirs
```

**Comment by vbraun** [comment]
I've seen it after `make doc-clean doc-uninstall` (because that is in my merge / incremental build script), I just don't have a repro. But it is out there. I run into this issue occasionally when making incremental builds.  

And the issue is that dependencies of docbuild artifacts are apparently insufficient. 

I think this branch exacerbates the issue because it places /doc/en/reference/spkg/index.rst under version control. So if you frequently change branches its easy to update timestaps which triggers incremental rebuilds, and then if you are unlucky docbuild trips over missing dependencies.

**Comment by kwankyu** [comment]
I just met the bug. As you say, It seems related with switching branches.

File: Makefile [hunk 1]
```diff
@@ -173,7 +173,7 @@ distclean: build-clean
 bootstrap-clean:
 	rm -rf config/install-sh config/compile config/config.guess config/config.sub config/missing configure build/make/Makefile-auto.in
 	rm -f src/doc/en/installation/*.txt
-	rm -rf src/doc/en/reference/spkg/*.rst
+	find src/doc/en/reference/spkg -name index.rst -prune -o -name *.rst -exec rm -f {} \+
 	for a in environment environment-optional src/environment src/environment-dev src/environment-optional; do rm -f $$a.yml $$a-3.[89].yml $$a-3.1[0-9].yml; done
```
Comment (kwankyu) [inline Makefile:177]: This line is the source of the "bug" Volker was experiencing.

Unlike previous `rm -rf src/doc/en/reference/spkg/*.rst`, where  the option `-r` is really ineffective, the `find` command searches the directory `src/doc/en/reference/spkg` and its subdirectories, and removes all `*.rst` files. In particular the file `src/doc/en/reference/spkg/sage/features.rst` is removed. 

The file is in control of the doc builder  and to be updated by it if necessary. Suppose that the file is created by the doc builder in other git branch. After switching to this PR branch, `./bootstrap` removes the file, but the doc builder still believes the file is intact, and the removed file is not recreated. Hence the boom. 

Incremental doc build mixed with switching git branches is inherently instable. But the changed behavior caused by this PR (the command in line 176) exacerbates the instability, as Volker could see through.

Comment (mkoeppe) [reply Makefile:177]: Thanks a lot for this detective work!

Comment (kwankyu) [reply Makefile:177]: Thanks. Works well with me (that is, no more the bug for me).

**Comment by mkoeppe** [comment]
Thanks!

**Pull request merged**